File: vendors/sainsburys.py
from datetime import datetime, timedelta, timezone
from urllib.parse import urlencode, quote

from playwright.async_api import Page

from .base import Cart, CartItem, Order, OrderDetail, OrderItem, Product, ShoppingVendor

_BASE = 'https://www.sainsburys.co.uk'
_BASKET = f'{_BASE}/groceries-api/gol-services/basket/v2/basket'
_PRODUCT = f'{_BASE}/groceries-api/gol-services/product/v1/product'
_ORDER = f'{_BASE}/groceries-api/gol-services/order/v1/order'

_ENABLED_FLAGS = (
    'add_to_favourites,use_food_basket_service,use_food_basket_service_v3,'
    'use_food_basket_service_v4,ads_conditionals,findability_v5,'
    'show_static_cnc_messaging,fetch_future_slot_weeks,'
    'click_and_collect_promo_banner,cookie_law_link,citrus_banners,'
    'citrus_favourites_trio_banners,offers_strategic_magnolia,special_logo,'
    'custom_product_messaging,promotional_link,promotional_link2,'
    'promotion_mechanics_page,findability_search,findability_autosuggest,'
    'fto_header_flag,recurring_slot_skip_opt_out,seasonal_favourites,'
    'cnc_start_amend_order_modal,favourites_product_cta_alt,'
    'get_favourites_from_v2,krang_alternatives,offers_config,'
    'alternatives_modal,relevancy_rank,changes_to_trolley,'
    'nectar_destination_page,unit_price_legislation,meal_deal_live,'
    'browse_pills_nav_type,use_cached_findability_results,event_zone_list,'
    'cms_carousel_zone_list,show_ynp_change_slot_banner,'
    'recipe_scrapbooks_enabled,event_carousel_skus,split_savings,'
    'trolley_nectar_card,favourites_magnolia,homepage,taggstar,'
    'meal_deal_cms_template_ids,pdp_accordions,pdp_occasions_pills,'
    'pdp_meta_desc_template,grouped_meal_deals,pci_phase_2,'
    'meal_deal_builder_nectar_widget,occasions_navigation,'
    'slots_event_banner_config,sales_window,resting_search,'
    'brands_background,brands_background_config,taggstar_config,'
    'all_ad_components_enabled,left_align_header,golui_my_addresses,'
    'new_global_header,new_filter_pages,spoonguru_disclaimers,'
    'recipe_reviews_enabled,sponsored_drawer,frequently_bought_together,'
    'show_ynp_opt_in_ui_elements,show_ynp_add_to_basket_toast,show_ynp_card,'
    'similar_products_drawer,fetch_ynp_opt_ins,resting_search_v2,bop_enabled,'
    'identity_transfer,prop_bar,favourites_boards,slot_confirmation_board,'
    'mobile_nav_2,highlight_seasonal_nav_item,should_not_scroll_into_view_fbt,'
    'show_popular_categories,compact_reviews,track_remove_scroll_experiment,'
    'favourites_grouped_by_top_category,track_boards_experiment,'
    'ynpoptin_national_launch,favourites_link_on_global_header,hey_sainsburys,'
    'heys_resting_state,krang_newness,show_tpr_straplines,'
    'track_compact_tile_experiment,track_pdp_occasions_pills_experiment,'
    'use_compact_tile_boards,use_compact_tile_previous_orders,use_compact_tile,'
    'occasions_closure_end_date_2025,favourites_view_all_AB_test,'
    'retry_your_payments,offers_revamp_2025_rollout,'
    'favourites_slot_your_usuals_tracking,product_bundles,fable_search_bar,'
    'hard_sku_replacement,track_occasions_available_from,continue_shopping_link,'
    'fto_first_available_slot,drip_pricing_phase_2_homepage,'
    'drip_pricing_phase_3_header,qualifying_basket_amount,app_banner,'
    'bigger_images,call_bcs,catchweight_dropdown,citrus_preview_new,'
    'citrus_search_trio_banners,citrus_xsell,compare_seasonal_favourites,'
    'constant_commerce_v2,ctt_ynp_products,desktop_interstitial_variant,'
    'disable_product_cache_validation,event_dates,favourites_pill_nav,'
    'favourites_whole_service,favourites_your_usuals_tracking,fbt_on_search,'
    'fbt_on_search_tracking,ff_abc_test_display,first_favourites_static,'
    'foodmaestro_modal,hfss_restricted,interstitial_variant,kg_price_label,'
    'krang_recommendations,lp_ab_test_display,lp_interstitial_grid_config,'
    'meal_planner,meganav,mobile_interstitial_variant,my_nectar_migration,'
    'nectar_card_associated,nectar_prices,new_favourites_filter,'
    'new_favourites_service,new_filters,ni_brexit_banner,occasions,'
    'offers_mechanics_carousel,optimised_product_tile,promo_lister_page,'
    'recipes_ingredients_modal,review_syndication,rokt,sale_january,'
    'search_cms,show_hd_xmas_slots_banner,similar_products,slot_v2,'
    'sponsored_featured_tiles,use_op_orchestrator_sde,xmas_dummy_skus,'
    'your_nectar_prices'
)

_FETCH_JS = """async ([method, url, body, headers]) => {
    const opts = { method, headers: { ...headers } };
    if (body !== null) {
        opts.headers['Content-Type'] = 'application/json';
        opts.body = JSON.stringify(body);
    }
    const r = await fetch(url, opts);
    const text = await r.text();
    let data;
    try { data = JSON.parse(text); } catch { data = text; }
    return { status: r.status, body: data };
}"""


class SainsburysVendor(ShoppingVendor):
    def __init__(self, page: Page):
        self._page = page
        self._access_token = ''
        self._wc_auth_token = ''
        self._store_identifier = '0652'
        self._flexi_stores = '0686'

    async def _init_context(self):
        ctx = await self._page.evaluate("""() => {
            const oidcRaw = localStorage.getItem('oidc.user:https://account.sainsburys.co.uk:gol');
            const oidc = oidcRaw ? JSON.parse(oidcRaw) : {};
            const wcMatch = document.cookie.match(/WC_AUTHENTICATION_\\d+=([^;]+)/);
            const slotRaw = localStorage.getItem('slot-reservation-cached');
            const slot = slotRaw ? JSON.parse(slotRaw) : {};
            return {
                accessToken: oidc.access_token || '',
                wcAuthToken: wcMatch ? wcMatch[1] : '',
                storeIdentifier: slot.storeIdentifier || '0652',
                flexiStores: (slot.flexiStores || [])[0] || '0686',
            };
        }""")
        self._access_token = ctx.get('accessToken') or ''
        self._wc_auth_token = ctx.get('wcAuthToken') or ''
        self._store_identifier = ctx.get('storeIdentifier') or '0652'
        self._flexi_stores = ctx.get('flexiStores') or '0686'

    def _pick_time(self) -> str:
        t = datetime.now(timezone.utc) + timedelta(days=1)
        return t.strftime('%Y-%m-%dT%H:%M:%SZ')

    def _headers(self) -> dict:
        return {
            'Authorization': f'Bearer {self._access_token}',
            'wcauthtoken': self._wc_auth_token,
            'enabled-feature-flags': _ENABLED_FLAGS,
            'accept': 'application/json',
        }

    async def _fetch(self, method: str, url: str, body=None) -> dict:
        result = await self._page.evaluate(_FETCH_JS, [method, url, body, self._headers()])
        if result['status'] == 401:
            await self._page.reload(wait_until='networkidle')
            await self._init_context()
            result = await self._page.evaluate(_FETCH_JS, [method, url, body, self._headers()])
        return result

    @staticmethod
    def _enc_cart_item(item_uid: str, product_uid: str, uom: str) -> str:
        return f'{item_uid}:{product_uid}:{uom}'

    @staticmethod
    def _dec_cart_item(cart_item_id: str) -> tuple[str, str, str]:
        item_uid, product_uid, uom = cart_item_id.split(':', 2)
        return item_uid, product_uid, uom

    def _basket_qs(self) -> str:
        return urlencode({
            'pick_time': self._pick_time(),
            'store_number': self._store_identifier,
            'slot_booked': 'false',
        })

    def _parse_cart(self, body: dict) -> Cart:
        items = []
        for i in body.get('items', []):
            product = i.get('product', {})
            items.append(CartItem(
                cart_item_id=self._enc_cart_item(
                    i['item_uid'], product.get('sku', ''), i.get('uom', 'ea')
                ),
                product_id=product.get('sku', ''),
                name=product.get('name', ''),
                qty=i.get('quantity', 0),
                price=f"£{i.get('subtotal_price', 0):.2f}",
            ))
        total = body.get('total_price', 0)
        savings = body.get('savings', 0)
        return Cart(
            items=items,
            total=f'£{total:.2f}',
            savings=f'£{abs(savings):.2f}' if savings else None,
        )

    async def search(self, term: str, size: int = 10) -> list[Product]:
        if not self._access_token:
            await self._init_context()
        qs = urlencode({
            'filter[keyword]': term,
            'page_number': 1,
            'page_size': size,
            'sort_order': 'FAVOURITES_FIRST',
            'store_identifier': self._store_identifier,
            'region': 'England',
            'flexi_stores': self._flexi_stores,
            'salesWindow': 1,
        })
        result = await self._fetch('GET', f'{_PRODUCT}?{qs}')
        if result['status'] != 200:
            raise RuntimeError(f'Search HTTP {result["status"]}')
        products = []
        for p in result['body'].get('products', []):
            retail = p.get('retail_price') or {}
            unit = p.get('unit_price') or {}
            promos = p.get('promotions') or []
            measure = unit.get('measure', '')
            size_str = f"{unit.get('measure_amount', 1)}{measure}" if measure and measure != 'ea' else None
            price_per = (f"£{unit['price']:.2f}/{measure}" if unit.get('price') and measure else None)
            products.append(Product(
                id=p['product_uid'],
                name=p.get('name', ''),
                size=size_str,
                price=f"£{retail.get('price', 0):.2f}",
                price_per_unit=price_per,
                promotion=promos[0].get('strap_line') if promos else None,
            ))
        return products

    async def get_cart(self) -> Cart:
        if not self._access_token:
            await self._init_context()
        result = await self._fetch('GET', f'{_BASKET}?{self._basket_qs()}')
        if result['status'] != 200:
            raise RuntimeError(f'Cart HTTP {result["status"]}')
        return self._parse_cart(result['body'])

    async def add(self, product_id: str, qty: int = 1) -> Cart:
        if not self._access_token:
            await self._init_context()
        body = {'product_uid': product_id, 'quantity': qty, 'uom': 'ea', 'selected_catchweight': ''}
        result = await self._fetch('POST', f'{_BASKET}/item?{self._basket_qs()}', body)
        if result['status'] not in (200, 201):
            raise RuntimeError(f'Add HTTP {result["status"]}')
        return self._parse_cart(result['body'])

    async def set_qty(self, cart_item_id: str, qty: int) -> Cart:
        if not self._access_token:
            await self._init_context()
        item_uid, product_uid, uom = self._dec_cart_item(cart_item_id)
        body = {'items': [{
            'product_uid': product_uid,
            'quantity': qty,
            'uom': uom,
            'selected_catchweight': '',
            'item_uid': item_uid,
        }]}
        result = await self._fetch('PUT', f'{_BASKET}?{self._basket_qs()}', body)
        if result['status'] != 200:
            raise RuntimeError(f'Set qty HTTP {result["status"]}')
        return self._parse_cart(result['body'])

    async def clear(self) -> Cart:
        if not self._access_token:
            await self._init_context()
        result = await self._fetch('DELETE', _BASKET)
        if result['status'] not in (200, 204):
            raise RuntimeError(f'Clear HTTP {result["status"]}')
        return await self.get_cart()

    async def get_orders(self, size: int = 15) -> list[Order]:
        if not self._access_token:
            await self._init_context()
        qs = urlencode({'page_size': size, 'page_number': 1})
        result = await self._fetch('GET', f'{_ORDER}?{qs}')
        if result['status'] != 200:
            raise RuntimeError(f'Orders HTTP {result["status"]}')
        orders = []
        for o in result['body'].get('orders', []):
            total = o.get('total')
            orders.append(Order(
                order_id=o['order_uid'],
                status=o.get('status', ''),
                placed_date=None,
                delivery_date=o.get('slot_start_time'),
                total=f'£{total:.2f}' if total is not None else None,
                item_count=None,
            ))
        return orders

    async def get_order(self, order_id: str) -> OrderDetail:
        if not self._access_token:
            await self._init_context()
        qs = urlencode({'placed': 'true', 'deliveryPass': 'false'})
        result = await self._fetch('GET', f'{_ORDER}/{order_id}?{qs}')
        if result['status'] != 200:
            raise RuntimeError(f'Order detail HTTP {result["status"]}')
        order = result['body']
        items = []
        for i in order.get('order_items', []):
            product = i.get('product', {})
            sub = i.get('sub_total')
            items.append(OrderItem(
                line_number=product.get('product_uid', ''),
                name=product.get('name'),
                size=None,
                qty=i.get('quantity'),
                unit_price=None,
                total_price=f'£{sub:.2f}' if sub is not None else None,
            ))
        total = order.get('total')
        return OrderDetail(
            order_id=order.get('order_uid', order_id),
            status=order.get('status', ''),
            placed_date=None,
            delivery_date=order.get('slot_start_time'),
            total=f'£{total:.2f}' if total is not None else None,
            items=items,
        )

File: packages/grocery-core/src/vendors/sainsburys.ts
import type { Page } from "playwright-core";
import type { Cart, CartItem, Order, OrderDetail, OrderItem, Product } from "../models.js";
import type { ShoppingVendor } from "./base.js";

type AnyRecord = Record<string, any>;

const BASE = "https://www.sainsburys.co.uk";
const BASKET = `${BASE}/groceries-api/gol-services/basket/v2/basket`;
const PRODUCT = `${BASE}/groceries-api/gol-services/product/v1/product`;
const ORDER = `${BASE}/groceries-api/gol-services/order/v1/order`;
const POUND = "\u00a3";

const ENABLED_FLAGS = [
  "add_to_favourites",
  "use_food_basket_service",
  "use_food_basket_service_v3",
  "use_food_basket_service_v4",
  "ads_conditionals",
  "findability_v5",
  "show_static_cnc_messaging",
  "fetch_future_slot_weeks",
  "click_and_collect_promo_banner",
  "cookie_law_link",
  "citrus_banners",
  "citrus_favourites_trio_banners",
  "offers_strategic_magnolia",
  "special_logo",
  "custom_product_messaging",
  "promotional_link",
  "promotion_mechanics_page",
  "findability_search",
  "findability_autosuggest",
  "fto_header_flag",
  "recurring_slot_skip_opt_out",
  "seasonal_favourites",
  "cnc_start_amend_order_modal",
  "favourites_product_cta_alt",
  "get_favourites_from_v2",
  "krang_alternatives",
  "offers_config",
  "alternatives_modal",
  "relevancy_rank",
  "changes_to_trolley",
  "nectar_destination_page",
  "unit_price_legislation",
  "meal_deal_live",
  "browse_pills_nav_type",
  "use_cached_findability_results",
  "event_zone_list",
  "cms_carousel_zone_list",
  "show_ynp_change_slot_banner",
  "recipe_scrapbooks_enabled",
  "event_carousel_skus",
  "split_savings",
  "trolley_nectar_card",
  "favourites_magnolia",
  "homepage",
  "taggstar",
  "meal_deal_cms_template_ids",
  "pdp_accordions",
  "pdp_occasions_pills",
  "pdp_meta_desc_template",
  "grouped_meal_deals",
  "pci_phase_2",
  "meal_deal_builder_nectar_widget",
  "occasions_navigation",
  "slots_event_banner_config",
  "sales_window",
  "resting_search",
  "brands_background",
  "taggstar_config",
  "all_ad_components_enabled",
  "left_align_header",
  "golui_my_addresses",
  "new_global_header",
  "new_filter_pages",
  "spoonguru_disclaimers",
  "recipe_reviews_enabled",
  "sponsored_drawer",
  "frequently_bought_together",
  "show_ynp_opt_in_ui_elements",
  "show_ynp_add_to_basket_toast",
  "show_ynp_card",
  "similar_products_drawer",
  "fetch_ynp_opt_ins",
  "resting_search_v2",
  "bop_enabled",
  "identity_transfer",
  "prop_bar",
  "favourites_boards",
  "slot_confirmation_board",
  "mobile_nav_2",
  "highlight_seasonal_nav_item",
  "show_popular_categories",
  "compact_reviews",
  "track_remove_scroll_experiment",
  "favourites_grouped_by_top_category",
  "track_boards_experiment",
  "ynpoptin_national_launch",
  "favourites_link_on_global_header",
  "hey_sainsburys",
  "heys_resting_state",
  "krang_newness",
  "show_tpr_straplines",
  "track_compact_tile_experiment",
  "use_compact_tile_boards",
  "use_compact_tile_previous_orders",
  "use_compact_tile",
  "retry_your_payments",
  "offers_revamp_2025_rollout",
  "favourites_slot_your_usuals_tracking",
  "product_bundles",
  "fable_search_bar",
  "hard_sku_replacement",
  "track_occasions_available_from",
  "continue_shopping_link",
  "fto_first_available_slot",
  "qualifying_basket_amount",
  "app_banner",
  "bigger_images",
  "call_bcs",
  "catchweight_dropdown",
  "citrus_preview_new",
  "citrus_search_trio_banners",
  "citrus_xsell",
  "compare_seasonal_favourites",
  "constant_commerce_v2",
  "ctt_ynp_products",
  "desktop_interstitial_variant",
  "disable_product_cache_validation",
  "event_dates",
  "favourites_pill_nav",
  "favourites_whole_service",
  "fbt_on_search",
  "foodmaestro_modal",
  "hfss_restricted",
  "krang_recommendations",
  "lp_ab_test_display",
  "lp_interstitial_grid_config",
  "meal_planner",
  "meganav",
  "mobile_interstitial_variant",
  "my_nectar_migration",
  "nectar_card_associated",
  "nectar_prices",
  "new_favourites_filter",
  "new_favourites_service",
  "new_filters",
  "ni_brexit_banner",
  "occasions",
  "offers_mechanics_carousel",
  "optimised_product_tile",
  "promo_lister_page",
  "recipes_ingredients_modal",
  "review_syndication",
  "rokt",
  "search_cms",
  "similar_products",
  "slot_v2",
  "sponsored_featured_tiles",
  "use_op_orchestrator_sde",
  "your_nectar_prices"
].join(",");

interface FetchResult extends Record<string, unknown> {
  status: number;
  body: any;
}

export function encodeSainsburysCartItemId(itemUid: string, productUid: string, uom: string): string {
  return `${itemUid}:${productUid}:${uom}`;
}

export function decodeSainsburysCartItemId(cartItemId: string): [string, string, string] {
  const [itemUid, productUid, uom] = cartItemId.split(":", 3);
  return [itemUid ?? "", productUid ?? "", uom ?? ""];
}

export class SainsburysVendor implements ShoppingVendor {
  private accessToken = "";
  private wcAuthToken = "";
  private storeIdentifier = "0652";
  private flexiStores = "0686";

  constructor(private readonly page: Page) {}

  static _enc_cart_item(itemUid: string, productUid: string, uom: string): string {
    return encodeSainsburysCartItemId(itemUid, productUid, uom);
  }

  static _dec_cart_item(cartItemId: string): [string, string, string] {
    return decodeSainsburysCartItemId(cartItemId);
  }

  async initContext(): Promise<void> {
    const ctx = await this.page.evaluate(() => {
      const oidcRaw = localStorage.getItem('oidc.user:https://account.sainsburys.co.uk:gol');
      const oidc = oidcRaw ? JSON.parse(oidcRaw) : {};
      const wcMatch = document.cookie.match(/WC_AUTHENTICATION_\d+=([^;]+)/);
      const slotRaw = localStorage.getItem('slot-reservation-cached');
      const slot = slotRaw ? JSON.parse(slotRaw) : {};
      return {
        accessToken: oidc.access_token || '',
        wcAuthToken: wcMatch ? wcMatch[1] : '',
        storeIdentifier: slot.storeIdentifier || '0652',
        flexiStores: (slot.flexiStores || [])[0] || '0686'
      };
    }) as AnyRecord;

    this.accessToken = String(ctx.accessToken ?? "");
    this.wcAuthToken = String(ctx.wcAuthToken ?? "");
    this.storeIdentifier = String(ctx.storeIdentifier ?? "0652");
    this.flexiStores = String(ctx.flexiStores ?? "0686");
  }

  async rawFetch(method: string, url: string, body: unknown = null): Promise<Record<string, unknown>> {
    return this.fetchJson(method, url, body);
  }

  async search(term: string, size = 10): Promise<Product[]> {
    if (!this.accessToken) {
      await this.initContext();
    }

    const qs = new URLSearchParams({
      "filter[keyword]": term,
      page_number: "1",
      page_size: String(size),
      sort_order: "FAVOURITES_FIRST",
      store_identifier: this.storeIdentifier,
      region: "England",
      flexi_stores: this.flexiStores,
      salesWindow: "1"
    });
    const result = await this.fetchJson("GET", `${PRODUCT}?${qs}`);
    if (result.status !== 200) {
      throw new Error(`Search HTTP ${result.status}`);
    }

    return arrayOfRecords(result.body?.products).map((product): Product => {
      const retail = product.retail_price ?? {};
      const unit = product.unit_price ?? {};
      const promos = arrayOfRecords(product.promotions);
      const measure = String(unit.measure ?? "");
      const sizeString = measure && measure !== "ea" ? `${unit.measure_amount ?? 1}${measure}` : null;
      const pricePer = unit.price && measure ? `${POUND}${Number(unit.price).toFixed(2)}/${measure}` : null;
      return {
        id: String(product.product_uid),
        name: String(product.name ?? ""),
        size: sizeString,
        price: money(retail.price),
        price_per_unit: pricePer,
        promotion: promos.length > 0 ? stringOrNull(promos[0].strap_line) : null
      };
    });
  }

  async getCart(): Promise<Cart> {
    if (!this.accessToken) {
      await this.initContext();
    }

    const result = await this.fetchJson("GET", `${BASKET}?${this.basketQs()}`);
    if (result.status !== 200) {
      throw new Error(`Cart HTTP ${result.status}`);
    }

    return this.parseCart(result.body ?? {});
  }

  async add(productId: string, qty = 1): Promise<Cart> {
    if (!this.accessToken) {
      await this.initContext();
    }

    const result = await this.fetchJson("POST", `${BASKET}/item?${this.basketQs()}`, {
      product_uid: productId,
      quantity: qty,
      uom: "ea",
      selected_catchweight: ""
    });
    if (![200, 201].includes(result.status)) {
      throw new Error(`Add HTTP ${result.status}`);
    }

    return this.parseCart(result.body ?? {});
  }

  async setQty(cartItemId: string, qty: number): Promise<Cart> {
    if (!this.accessToken) {
      await this.initContext();
    }

    const [itemUid, productUid, uom] = decodeSainsburysCartItemId(cartItemId);
    const result = await this.fetchJson("PUT", `${BASKET}?${this.basketQs()}`, {
      items: [{
        product_uid: productUid,
        quantity: qty,
        uom,
        selected_catchweight: "",
        item_uid: itemUid
      }]
    });
    if (result.status !== 200) {
      throw new Error(`Set qty HTTP ${result.status}`);
    }

    return this.parseCart(result.body ?? {});
  }

  async clear(): Promise<Cart> {
    if (!this.accessToken) {
      await this.initContext();
    }

    const result = await this.fetchJson("DELETE", BASKET);
    if (![200, 204].includes(result.status)) {
      throw new Error(`Clear HTTP ${result.status}`);
    }

    return this.getCart();
  }

  async getOrders(size = 15): Promise<Order[]> {
    if (!this.accessToken) {
      await this.initContext();
    }

    const qs = new URLSearchParams({ page_size: String(size), page_number: "1" });
    const result = await this.fetchJson("GET", `${ORDER}?${qs}`);
    if (result.status !== 200) {
      throw new Error(`Orders HTTP ${result.status}`);
    }

    return arrayOfRecords(result.body?.orders).map((order): Order => {
      const total = order.total;
      return {
        order_id: String(order.order_uid),
        status: String(order.status ?? ""),
        placed_date: null,
        delivery_date: stringOrNull(order.slot_start_time),
        total: total !== undefined && total !== null ? money(total) : null,
        item_count: null
      };
    });
  }

  async getOrder(orderId: string): Promise<OrderDetail> {
    if (!this.accessToken) {
      await this.initContext();
    }

    const qs = new URLSearchParams({ placed: "true", deliveryPass: "false" });
    const result = await this.fetchJson("GET", `${ORDER}/${orderId}?${qs}`);
    if (result.status !== 200) {
      throw new Error(`Order detail HTTP ${result.status}`);
    }

    const order = result.body ?? {};
    const items: OrderItem[] = arrayOfRecords(order.order_items).map((item) => {
      const product = item.product ?? {};
      const subtotal = item.sub_total;
      return {
        product_id: idOrNull(product.product_uid),
        name: stringOrNull(product.name),
        size: null,
        qty: numberOrNull(item.quantity),
        unit_price: null,
        total_price: subtotal !== undefined && subtotal !== null ? money(subtotal) : null
      };
    });

    const total = order.total;
    return {
      order_id: String(order.order_uid ?? orderId),
      status: String(order.status ?? ""),
      placed_date: null,
      delivery_date: stringOrNull(order.slot_start_time),
      total: total !== undefined && total !== null ? money(total) : null,
      items
    };
  }

  private async fetchJson(method: string, url: string, body: unknown = null): Promise<FetchResult> {
    const request: [string, string, unknown, Record<string, string>] = [method, url, body ?? null, this.headers()];
    let result = await this.page.evaluate(fetchInPage, request) as FetchResult;
    if (result.status === 401) {
      await this.page.reload({ waitUntil: "networkidle" });
      await this.initContext();
      const retryRequest: [string, string, unknown, Record<string, string>] = [method, url, body ?? null, this.headers()];
      result = await this.page.evaluate(fetchInPage, retryRequest) as FetchResult;
    }

    return result;
  }

  private headers(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.accessToken}`,
      wcauthtoken: this.wcAuthToken,
      "enabled-feature-flags": ENABLED_FLAGS,
      accept: "application/json"
    };
  }

  private basketQs(): string {
    return new URLSearchParams({
      pick_time: this.pickTime(),
      store_number: this.storeIdentifier,
      slot_booked: "false"
    }).toString();
  }

  private pickTime(): string {
    const date = new Date(Date.now() + 24 * 60 * 60 * 1000);
    return date.toISOString().replace(/\.\d{3}Z$/, "Z");
  }

  private parseCart(body: AnyRecord): Cart {
    const items: CartItem[] = arrayOfRecords(body.items).map((item) => {
      const product = item.product ?? {};
      return {
        cart_item_id: encodeSainsburysCartItemId(
          String(item.item_uid),
          String(product.sku ?? ""),
          String(item.uom ?? "ea")
        ),
        product_id: String(product.sku ?? ""),
        name: String(product.name ?? ""),
        qty: Number(item.quantity ?? 0),
        price: money(item.subtotal_price)
      };
    });

    const total = Number(body.total_price ?? 0);
    const savings = Number(body.savings ?? 0);
    return {
      items,
      total: money(total),
      savings: savings ? money(Math.abs(savings)) : null
    };
  }
}

function arrayOfRecords(value: unknown): AnyRecord[] {
  return Array.isArray(value) ? value.filter((entry) => entry && typeof entry === "object") as AnyRecord[] : [];
}

function stringOrNull(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function idOrNull(value: unknown): string | null {
  return value === undefined || value === null || value === "" ? null : String(value);
}

function numberOrNull(value: unknown): number | null {
  return typeof value === "number" ? value : null;
}

function money(value: unknown): string {
  return `${POUND}${Number(value ?? 0).toFixed(2)}`;
}

async function fetchInPage(
  [method, url, body, headers]: [string, string, unknown, Record<string, string>]
): Promise<FetchResult> {
  const opts: RequestInit = {
    method,
    headers: { ...headers }
  };
  if (body !== null) {
    (opts.headers as Record<string, string>)["Content-Type"] = "application/json";
    opts.body = JSON.stringify(body);
  }

  const response = await fetch(url, opts);
  const text = await response.text();
  let data: unknown;
  try {
    data = JSON.parse(text);
  } catch {
    data = text;
  }

  return { status: response.status, body: data };
}
